vm/src/stdlib/ctypes/basics.rs

```diff
@@ -0,0 +1,507 @@
+                                Err(vm.new_type_error("'_length_' must positive".to_string()))
+                            } else {
+                                Ok(v_int as usize)
+                            }
+                        }
+                        _ => Err(vm.new_type_error("'_length_' must be an integer".to_string())),
+                    })?;
+
+                Ok(RawBuffer {
+                    inner: buf as *const u8 as *mut _,
+                    size: len,
+                })
+            }
+            Ok(_) => Err(vm.new_type_error("abstract class".to_string())),
+            // @TODO: A sanity check
+            Err(_) => Err(vm.new_type_error("attribute '__abstract__' must be bool".to_string())),
+        },
+        // @TODO: I think it's unreachable
+        Err(_) => Err(vm.new_attribute_error("abstract class".to_string())),
+    }
+}
+
+fn buffer_copy(
+    cls: PyTypeRef,
+    obj: PyObjectRef,
+    offset: OptionalArg,
+    vm: &VirtualMachine,
+    copy: bool,
+) -> PyResult<PyCData> {
+    match vm.get_attribute(cls.as_object().to_owned(), "__abstract__") {
+        Ok(attr) => {
+            match bool::try_from_object(vm, attr) {
+                Ok(b) if !b => {
+                    let buffer = try_buffer_from_object(vm, &obj)?;
+                    let opts = buffer.get_options().clone();
+
+                    // @TODO: Fix the way the size of stored
+                    // Would this be the a proper replacement?
+                    // vm.call_method(cls.as_object().to_owned(), "size", ())?.
+                    let cls_size = vm
+                        .get_attribute(cls.as_object().to_owned(), "_size")
+                        .map(|c_s| usize::try_from_object(vm, c_s))??;
+
+                    let offset_int = offset
+                        .into_option()
+                        .map_or(Ok(0), |off| i64::try_from_object(vm, off))?;
+
+                    if opts.readonly {
+                        Err(vm.new_type_error("underlying buffer is not writable".to_string()))
+                    } else if !opts.contiguous {
+                        Err(vm.new_type_error("underlying buffer is not C contiguous".to_string()))
+                    } else if offset_int < 0 {
+                        Err(vm.new_value_error("offset cannot be negative".to_string()))
+                    } else if cls_size > opts.len - (offset_int as usize) {
+                        Err(vm.new_value_error(format!(
+                            "Buffer size too small ({} instead of at least {} bytes)",
+                            cls_size,
+                            opts.len + (offset_int as usize)
+                        )))
+                    } else if let Some(mut buffer) = buffer.as_contiguous_mut() {
+                        // @TODO: Is this copying?
+
+                        let buffered = if copy {
+                            unsafe { slice::from_raw_parts_mut(buffer.as_mut_ptr(), buffer.len()) }
```

Comment:
this is not a copy

Reply:
Here
```rust
let buffered = if copy {
    unsafe { slice::from_raw_parts_mut(buffer.as_mut_ptr(), buffer.len()) }
        .as_mut_ptr()
} else {
    buffer.as_mut_ptr()
};
```
The idea is to copy the bytes from the buffer if `copy`, or return a view otherwise. Should I use `to_contiguous` to copy it?